### PR TITLE
Make aggregations in INSERT INTO work

### PR DESCRIPTION
**Pull request by mkleen**
Aggregations used in `INSERT INTO` statement will return null e.g.:

```sql
INSERT INTO t2 SELECT AVG(a) FROM t1;
```

This applies to all types of aggregations used in this context. The root cause is that references in the ``INSERT INTO`` have a hint to prefer source look-ups, which should be ignored by aggregations as it leads to the wrong field name for doc values.

Fixes https://github.com/crate/crate/issues/14277

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java
Comment (mkleen) [inline extensions/functions/src/main/java/io/crate/operation/aggregation/HyperLogLogDistinctAggregation.java:276]: Changing to `ColumnIdent` reduces the chances of using an invalid column name.

File: docs/appendices/release-notes/unreleased.rst
Comment (seut) [inline docs/appendices/release-notes/unreleased.rst:201]: Is this a regression introduced with recent releases? If so, lets point that out please.

File: server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java
Comment (seut) [inline server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:164]: Can we elaborate why it don't make sense? Like, I think it not only make no real sense but break things!

File: server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:1853]:
```suggestion
    public void test_insert_into_using_aggregations() throws InterruptedException {
```

Comment (seut) [inline server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:1853]: We already have a similar test `testInsertFromQueryGlobalAggregate`, I wonder why this doesn't fail?

Comment (mkleen) [reply server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:1853]: This breaks as soon as you remove the `count(*)` with the same failure (null) e.g.:
``insert into aggs (c, s) (select sum(position), sum(position) from locations)``

Comment (mkleen) [reply server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:1853]: We basically only have the problem `PlanHint.PREFER_SOURCE_LOOKUP` comes into play.

File: server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java
Comment (mkleen) [reply server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:164]: The field does not exist in doc values. I will add it to the comment.

File: docs/appendices/release-notes/unreleased.rst
Comment (matriv) [inline docs/appendices/release-notes/unreleased.rst:200]: minor: 
```suggestion
- Fixed an issue which caused aggregations used in ``INSERT INTO`` statements
```

File: server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:167]: I think switching over to regular references would be worthwhile for aggregations even outside of the `DocValuesAggregates`.

Could be done in the planner, similiar to https://github.com/crate/crate/pull/11498/files#diff-9ca8b12f552725efbcc105a2530e9e6d74ae7f91dcb545c95c30132219b59f95 

(And maybe also check the performance impact, https://github.com/crate/crate/pull/11498 had quite an improvement9

Comment (mkleen) [reply server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:167]: Thanks for the feedback.

File: server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:566]: 1. count is also an aggregation but was not failing. Do I get it right, that it's kinda exceptional and could work with source lookup?
2. since we updated this test, I wonder if we need a new one then. no strong opinion though.

**Review by BaurzhanSakhariev — APPROVED** [review]
LGTM, added some questions. 

Please wait for the approval from others.

**Comment by mkleen** [comment]
Benchmarks on the way.

File: server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java
Comment (mkleen) [reply server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:566]: Good point, i think one is enough.

**Review by mfussenegger — COMMENTED** [review]
Could you add an assertion in `DocValuesAggregates` that the references are not starting with `_doc`?

File: server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java
Comment (seut) [inline server/src/main/java/io/crate/execution/engine/collect/DocValuesAggregates.java:163]:
```suggestion
                    assert reference.ident().columnIdent().fqn().startsWith(DocSysColumns.Names.DOC) == false :
```

**Comment by BaurzhanSakhariev** [comment]
I think we don't do 5.2 hotfixes anymore?

File: server/src/main/java/io/crate/planner/operators/HashAggregate.java
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/operators/HashAggregate.java:80]: The primary reason here should be performance.  Global aggregations are a pipeline breaker using all data.  So column store is likely more efficient. 

That the implementation (currently) also can't use source lookups is secondary as its an implementation detail that shouldn't matter as much for the planner

(Hence also the assertion for the implementation.  It's not ideal that one far removed component relies on another component acting in a certain way,  but I think the trade-off is acceptable here. Otherwise we'd need to add the rewrite to non-source lookup you had in the initial version back too - only for it to be dead coded)

Comment (mkleen) [reply server/src/main/java/io/crate/planner/operators/HashAggregate.java:80]: ok 👍

**Comment by mkleen** [comment]
Aggregation benchmarks seem to be uneffected.

```
V1: 5.4.0-f006fb5d95c8256639ab7655138fe500a9ae9473 -> (master)
V2: 5.4.0-963f3f9ef1cfd98aa9106f2805cc86a18fad4c31 -> (mkleen/fix-aggregation-references )


Q: select avg("adRevenue") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       16.958 ±   40.180 |     12.886 |     14.182 |     15.545 |    909.464 |
|   V2    |       17.102 ±   42.495 |     12.765 |     14.184 |     15.295 |    958.934 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   0.84%                           +   0.01%
There is a 4.39% probability that the observed difference is not random, and the best estimate of that difference is 0.84%
The test has no statistical significance

Q: select max(duration) from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       14.650 ±    3.441 |     13.117 |     14.183 |     14.785 |     84.915 |
|   V2    |       14.627 ±    2.765 |     13.181 |     14.192 |     14.649 |     66.868 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   0.16%                           +   0.07%
There is a 9.59% probability that the observed difference is not random, and the best estimate of that difference is 0.16%
The test has no statistical significance

Q: select max("adRevenue") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       13.781 ±    3.005 |     12.805 |     13.326 |     13.726 |     74.861 |
|   V2    |       13.832 ±    3.254 |     13.005 |     13.417 |     13.643 |     82.412 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   0.37%                           +   0.68%
There is a 20.09% probability that the observed difference is not random, and the best estimate of that difference is 0.37%
The test has no statistical significance

Q: select min(duration) from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       12.352 ±    1.270 |     11.416 |     12.124 |     12.447 |     30.295 |
|   V2    |       12.250 ±    1.094 |     11.623 |     12.025 |     12.241 |     26.267 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   0.83%                           -   0.82%
There is a 82.77% probability that the observed difference is not random, and the best estimate of that difference is 0.83%
The test has no statistical significance

Q: select min("adRevenue") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       13.341 ±    1.017 |     12.630 |     13.141 |     13.453 |     33.046 |
|   V2    |       14.860 ±    0.919 |     14.315 |     14.676 |     14.895 |     27.440 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +  10.77%                           +  11.04%
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 10.77%
The test has statistical significance

Q: select max_by("adRevenue", "visitDate") from uservisits
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       12.959 ±    2.917 |     11.935 |     12.422 |     12.943 |     61.771 |
|   V2    |       13.023 ±    2.529 |     12.112 |     12.625 |     13.002 |     66.306 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   0.49%                           +   1.63%
There is a 28.82% probability that the observed difference is not random, and the best estimate of that difference is 0.49%
The test has no statistical significance

Q: select sum(i) from (select '2 hours'::interval from generate_series(1, 100000)) as t (i)
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       17.689 ±    8.092 |     15.409 |     16.540 |     17.149 |    174.652 |
|   V2    |       17.544 ±    6.962 |     15.485 |     16.485 |     17.215 |    147.193 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -   0.82%                           -   0.34%
There is a 23.81% probability that the observed difference is not random, and the best estimate of that difference is 0.82%
The test has no statistical significance
```

**Review by mfussenegger — APPROVED** [review]

**Comment by mkleen** [comment]
> > Aggregation benchmarks seem to be uneffected.
> 
> You compared something that this PR isn't changing. It needs to include operators which set the source lookup hint (e.g. inserts), and it needs to use aggregations in a way that wasn't broken - by bypassing the doc-value-aggregates.
> 
> For example:
> 
> ```toml
> [setup]
> statements = [
>   "create table tsrc (x long)",
>   "create table tdst (x long)",
>   "insert into tsrc (x) (select x from generate_series(1, 1_000_000) t (x))",
>   "refresh table tsrc"
> ]
> 
> 
> [[queries]]
> name = "insert"
> statement = "insert into tdst (x) (select avg(x) filter (where (x % 2 = 0)) from tsrc)"
> iterations = 1
> 
> 
> [teardown]
> statements = [
>   "drop table if exists tsrc",
>   "drop table if exists tdst",
> ]
> ```
> 
> Without this change:
> 
> ```
> Runtime (in ms):
>     mean:    1064.876 ± 0.000
> # Running tearDown
> ```
> 
> With this change:
> 
> ```
> Runtime (in ms):
>     mean:    170.587 ± 0.000
> # Running tearDown
> ```
> 
> So it does improve performance. This is why I mentioned that the change likely also makes sense for the general aggregation code path, not just for the `DocValueAggregates`

Thanks for that. I understood that I should just check normal aggregations.

**Pull request merged**
